WordPress/WordPressApi/WordPressComOAuthClient.m

```diff
@@ -72,36 +92,45 @@ - (void)requestOneTimeCodeWithUsername:(NSString *)username
-       success:^(AFHTTPRequestOperation *operation, id responseObject) {
-           if (success) {
-               success();
-           }
-       } failure:^(AFHTTPRequestOperation *operation, NSError *error) {
-           error = [self processError:error forOperation:operation];
-           
-           // SORRY:
-           // SMS Requests will still return WordPressComOAuthErrorNeedsMultifactorCode. In which case,
-           // we should hit the success callback.
-           if (error.code == WordPressComOAuthErrorNeedsMultifactorCode) {
-               if (success) {
-                   success();
-               }
-           } else if (failure) {
-               failure(error);
-           }
-       }];
+    [self.sessionManager POST:@"token"
+                   parameters:parameters
+                      success:^(NSURLSessionDataTask *task, id responseObject) {
+                          if (success) {
+                              success();
+                          }
+                      } failure:^(NSURLSessionDataTask *task, NSError *error) {
+                          if ([task.response isKindOfClass:[NSHTTPURLResponse class]]) {
+                              NSHTTPURLResponse *httpURLResponse = (NSHTTPURLResponse *)task.response;
+                              error = [self processError:error forResponse:httpURLResponse];
+                          }
+
+                          // SORRY:
+                          // SMS Requests will still return WordPressComOAuthErrorNeedsMultifactorCode. In which case,
+                          // we should hit the success callback.
+                          if (error.code == WordPressComOAuthErrorNeedsMultifactorCode) {
+                              if (success) {
+                                  success();
+                              }
+                          } else if (failure) {
+                              failure(error);
+                          }
+                      }];
 }
 
-- (NSError *)processError:(NSError *)error forOperation:(AFHTTPRequestOperation *)operation {
-    if (operation.response.statusCode >= 400 && operation.response.statusCode < 500) {
+- (NSError *)processError:(NSError *)error forResponse:(NSHTTPURLResponse *)response {
+    if (response.statusCode >= 400 && response.statusCode < 500) {
         // Bad request, look for errors in the JSON response
-		NSDictionary* response = operation.responseObject;
 
-        if (response) {
-            NSString *errorCode = [response stringForKey:@"error"];
-            NSString *errorDescription = [response stringForKey:@"error_description"];
+        NSData* responseData = nil;
+        NSDictionary *responseDictionary = nil;
+        if (error.domain == AFURLResponseSerializationErrorDomain) {
+            responseData = error.userInfo[AFNetworkingOperationFailingURLResponseDataErrorKey];
+            responseDictionary = [NSJSONSerialization JSONObjectWithData:responseData options:nil error:nil];
```

Comment:
The thing is that because the server returns status code of between 400 and 500 AFNetworking will consider it a failure and don't even try to parse the answer... it will simple return on the failure callback... so we need to go and check if there is any detail there to show to the user.